hw/src/ledger.rs

```diff
@@ -0,0 +1,364 @@
+	fn from(err: hidapi::HidError) -> Error {
+		Error::Usb(err)
+	}
+}
+
+/// Ledger device manager.
+pub struct Manager {
+	usb: hidapi::HidApi,
+	devices: Vec<Device>,
+	key_path: KeyPath,
+}
+
+#[derive(Debug)]
+struct Device {
+	path: String,
+	info: WalletInfo,
+}
+
+impl Manager {
+	/// Create a new instance.
+	pub fn new() -> Result<Manager, Error> {
+		let manager = Manager {
+			usb: hidapi::HidApi::new()?,
+			devices: Vec::new(),
+			key_path: KeyPath::Ethereum,
+		};
+		Ok(manager)
+	}
+
+	/// Re-populate device list. Only those devices that have Ethereum app open will be added.
+	pub fn update_devices(&mut self) -> Result<usize, Error> {
+		self.usb.refresh_devices();
+		let devices = self.usb.devices();
+		let mut new_devices = Vec::new();
+		let mut num_new_devices = 0;
+		for device in devices {
+			trace!("Checking device: {:?}", device);
+			if device.vendor_id != LEDGER_VID || !LEDGER_PIDS.contains(&device.product_id) {
+				continue;
+			}
+			match self.read_device_info(&device) {
+				Ok(info) => {
+					debug!("Found device: {:?}", info);
+					if !self.devices.iter().any(|d| d.path == info.path) {
+						num_new_devices += 1;
+					}
+					new_devices.push(info);
+
+				},
+				Err(e) => debug!("Error reading device info: {}", e),
+			};
+		}
+		self.devices = new_devices;
+		Ok(num_new_devices)
+	}
+
+	/// Select key derivation path for a known chain.
+	pub fn set_key_path(&mut self, key_path: KeyPath) {
+		self.key_path = key_path;
+	}
+
+	fn read_device_info(&self, dev_info: &hidapi::HidDeviceInfo) -> Result<Device, Error> {
+		let mut handle = self.open_path(&dev_info.path)?;
+		let address = Self::read_wallet_address(&mut handle, self.key_path)?;
+		let manufacturer = dev_info.manufacturer_string.clone().unwrap_or("Unknown".to_owned());
+		let name = dev_info.product_string.clone().unwrap_or("Unknown".to_owned());
+		let serial = dev_info.serial_number.clone().unwrap_or("Unknown".to_owned());
+		Ok(Device {
+			path: dev_info.path.clone(),
+			info: WalletInfo {
+				name: name,
+				manufacturer: manufacturer,
+				serial: serial,
+				address: address,
+			},
+		})
+	}
+
+	fn read_wallet_address(handle: &hidapi::HidDevice, key_path: KeyPath) -> Result<Address, Error> {
+		let ver = Self::send_apdu(handle, commands::GET_APP_CONFIGURATION, 0, 0, &[])?;
+		if ver.len() != 4 {
+			return Err(Error::Protocol("Version packet size mismatch"));
+		}
+
+		let (major, minor, patch) = (ver[1], ver[2], ver[3]);
+		if major <= 1 && minor == 0 && patch < 3 {
+			return Err(Error::Protocol("App version 1.0.3 is required."));
+		}
+
+		let eth_path = &ETH_DERIVATION_PATH_BE[..];
+		let etc_path = &ETC_DERIVATION_PATH_BE[..];
+		let derivation_path = match key_path {
+			KeyPath::Ethereum => eth_path,
+			KeyPath::EthereumClassic => etc_path,
+		};
+		let key_and_address = Self::send_apdu(handle, commands::GET_ETH_PUBLIC_ADDRESS, 0, 0, derivation_path)?;
+		if key_and_address.len() != 107 { // 1 + 65 PK + 1 + 20 Addr
```

Comment:
Address actually takes 40 because it comes ascii-hex encoded. Will update the comment.